### PR TITLE
feat: tls access relation

### PR DESCRIPTION
**Pull request by kayra1**
# Description

This change introduces the notary tls requires endpoint for certificate providers to provide an access certificate.

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that validate the behaviour of the software
- [x] I validated that new and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/integration/test_charm.py [hunk 1]
```diff
@@ -43,9 +44,66 @@ async def test_build_and_deploy(ops_test: OpsTest, request: pytest.FixtureReques
...
+        )
+    new_ca = await get_file_from_notary(ops_test, "ca.pem")
+    assert new_ca != first_ca
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME, TLS_PROVIDER_APPLICATION_NAME],
+        status="active",
+        timeout=1000,
+        raise_on_error=True,
+    )
+
```
Comment (DanielArndt) [inline tests/integration/test_charm.py:105]: Why is this necessary?

File: tests/integration/test_charm.py [hunk 2]
```diff
@@ -43,9 +44,66 @@ async def test_build_and_deploy(ops_test: OpsTest, request: pytest.FixtureReques
...
+        trust=True,
+    )
+    await ops_test.model.deploy(
+        "loki-k8s", application_name=LOKI_APPLICATION_NAME, trust=True, channel="stable"
+    )
     await ops_test.model.wait_for_idle(apps=[APP_NAME], status="active", timeout=1000)
 
 
+@pytest.mark.abort_on_fail
+async def test_given_tls_access_relation_when_related_and_unrelated_to_notary_then_certificates_replaced_appropriately(
```
Comment (DanielArndt) [inline tests/integration/test_charm.py:71]: Where is the `and_unrelated` part of this test?

File: tests/unit/test_charm.py [hunk 3]
```diff
@@ -3395,10 +3417,180 @@ def test_given_certificate_rejected_in_notary_when_configure_then_certificate_re
...
             ),
         ):
             context.run(context.on.update_status(), state)
         mock_set_relation_certificate.assert_called_once()
+
+    @patch(f"{TLS_LIB_PATH}.TLSCertificatesRequiresV4.get_assigned_certificate")
+    def test_given_access_relation_created_when_configure_then_certificate_not_replaced(
+        self, mock_assigned_certificates, context
+    ):
+        with tempfile.TemporaryDirectory() as tempdir:
```
Comment (DanielArndt) [inline tests/unit/test_charm.py:3434]: I believe it is preferable to use pytest's [`tmp_path`](https://docs.pytest.org/en/stable/how-to/tmp_path.html).

File: tests/unit/test_charm.py [hunk 4]
```diff
@@ -3395,10 +3417,180 @@ def test_given_certificate_rejected_in_notary_when_configure_then_certificate_re
...
+                ],
+                relations=[Relation(id=1, endpoint=TLS_ACCESS_RELATION_NAME)],
+                leader=True,
+            )
+            certificate, _ = self.example_cert_and_key()
+            with open(tempdir + "/certificate.pem", "w") as f:
+                f.write(str(certificate))
+            mock_assigned_certificates.return_value = (None, None)
+            with patch(
+                "notary.Notary.__new__",
```
Comment (DanielArndt) [inline tests/unit/test_charm.py:3469]: I think I remember reading somewhere that patching `__new__` is dangerous and can lead to some unexpected / undefined behaviour.

I was working around this by patching the class (`notary.Notary`) with a mock that returned another mock (i.e. it had a `return_value` that was a mock itself). I admit it isn't pretty, but that is what the pytest authors seemed to encourage. I don't know where I read that though, so do with that information what you'd like!

**Review by DanielArndt — COMMENTED** [review]
Looks good to me,  a few minor comments.

File: tests/integration/test_charm.py (hunk 2 above)
Comment (kayra1) [reply tests/integration/test_charm.py:71]: Added back

File: tests/integration/test_charm.py (hunk 1 above)
Comment (kayra1) [reply tests/integration/test_charm.py:105]: It is probably not. removed

File: tests/unit/test_charm.py (hunk 4 above)
Comment (kayra1) [reply tests/unit/test_charm.py:3469]: I will choose to ignore it, because it took me a ton of time to get this to work :D

File: tests/unit/test_charm.py (hunk 3 above)
Comment (kayra1) [reply tests/unit/test_charm.py]: done

File: tests/integration/test_charm.py [hunk 5]
```diff
@@ -58,16 +126,10 @@ async def test_given_notary_when_tls_requirer_related_then_csr_uploaded_to_notar
...
-        "tls-certificates-requirer",
-        application_name=TLS_REQUIRER_APPLICATION_NAME,
-        channel="edge",
-        trust=True,
-    )
     await ops_test.model.integrate(
         relation1=f"{APP_NAME}:certificates",
         relation2=f"{TLS_REQUIRER_APPLICATION_NAME}",
-    )
+    )  # type: ignore
```
Comment (DanielArndt) [inline tests/integration/test_charm.py:132]: not sure why this was needed?

**Review by DanielArndt — APPROVED** [review]

File: tests/integration/test_charm.py (hunk 5 above)
Comment (kayra1) [reply tests/integration/test_charm.py]: looks like it wasnt. gone
